runtime/major_gc.c

```diff
@@ -939,33 +939,39 @@ Caml_noinline static intnat do_some_marking(struct mark_stack* stk,
   return budget;
 }
 
-/* compressed mark stack */
-#define PAGE_MASK (~(uintnat)(BITS_PER_WORD-1))
-#define PTR_TO_PAGE(v) (((uintnat)(v)/sizeof(value)) & PAGE_MASK)
-#define PTR_TO_PAGE_OFFSET(v) ((((uintnat)(v)/sizeof(value)) & ~PAGE_MASK))
+/* Compressed mark stack
+
+   We use a bitset, implemented as a hashtable storing word-sized
+   integers (uintnat). Each integer represents a "chunk" of addresses
+   that may or may not be present in the stack.
+ */
+#define CHUNK_MASK (~(uintnat)(BITS_PER_WORD-1))
+#define PTR_TO_CHUNK(v) (((uintnat)(v)/sizeof(value)) & CHUNK_MASK)
+#define PTR_TO_CHUNK_OFFSET(v) ((((uintnat)(v)/sizeof(value)) & ~CHUNK_MASK))
+#define CHUNK_AND_OFFSET_TO_PTR(chunk, offset) \
+  ((value*) ((chunk + offset) * sizeof(value)))
```

Comment:
I would guess that the c compilers that end up getting used for this c code would do well with the following definitions, which get rid of the macros. They should end up generating the same code.

```suggestion
static const uintnat chunk_mask = ~(uintnat)(BITS_PER_WORD-1);
static inline uintnat ptr_to_chunk(value v){ return ((uintnat)(v)/sizeof(value)) & chunk_mask; }
static inline uintnat ptr_to_chunk_offset(value v){ return ((uintnat)(v)/sizeof(value)) & ~chunk_mask; }
static inline value* chunk_and_offset_to_ptr(uintnat chunk, uintnat offset){
  return  (value*) ((chunk + offset) * sizeof(value));
}
```

Reply:
I vaguely thought of using functions instead of macros. One advantage of the proposal that only became apparent when I saw your patch (duh!) is that it avoids the namespacing issues of macros entirely.